### DISCUSSION TITLE
Working with socket IO on clustered environment ( NodeJS clusters)

### DISCUSSION THREAD

**Discussion by rituraj12797** [socketio/socket.io#4983] [Q&A] [answered]
I am operating on a clustered environment using NodeJS clusters , I am trying to send direct messages from a client in worker/ room 1 to a client in worker/room `n`  what initially i was doing is that

a. i will transmit the message with the data of the room of recipient and the socketID of recipient
but the problem was that when i was sending this message to the room via the IPC provided by cluster adapter
the message is broadcasted to al the clients on reaching the room due to the code

   `socket.to(`${room_no}1).emit("message","<some_message>)` 
   
b. eventually i figured out that somehow just by writing this code

 socket.to(${socketId}).emit("message", {text});

it would automatically search for the client with the given socketID globally ( idk how is this working if anyone could help )
and automatically transmits the message from the sender (in room a) to the receiver (in room b) eliminating the need of finding the room.

still i would like to know if anyone could help me finding how is it able to globally search for the recipient how efficient is it

**Comment by darrachequesne** [comment]
> idk how is this working if anyone could help

If you use the `@socket.io/cluster-adapter` package, when calling `socket.to(socketId).emit("hello", "world")` the server will send the packet to the other workers, including the targeted rooms, something like:

```js
{
  data: {
    packet: {
      data: ["hello", "world"]
    },
    opts: {
      rooms: ["theSocketId"]
    }
  }
}
```

Then each worker will loop through their own sockets and find the one that is in the room named `socketId`.

**Reply by rituraj12797** [reply]
So all it's doing is brute force linear search among all the clients for all the workers to find the appropriate one. 

What I was initially trying to do was that i would maintain a hash table when a client connects to a worker I would store the (WorkerId, clientId) in hash table using redis

Now when I will send the message through the IPC channel, I will first fetch the worker ID where I want to send using the socket id(client ID) from the hash table, to find the room in constant time, then i will send the data to that worker only instead of a sending to every worker and increased traffic in IPC channel. 

Now that I am in the correct room I will search for the matching client there and send it, reducing the time. 

Is there any way I could implement the above approach? 
and please do correct me anywhere I am wrong.

**Reply by darrachequesne** [reply, marked as answer]
In that case, I guess you could store the relationship in the primary process, and then forward the message to the right worker here: https://github.com/socketio/socket.io-cluster-adapter/blob/56a53bceb9a0f3032c68e9a9ae63643423890cb1/lib/index.ts#L552-L579

**Comment by rituraj12797** [comment]
When using hash-based routing with `Socket.to(room_id)`, the message goes to every client in the room, causing unnecessary traffic. Meanwhile, `io.to(socket_id)`  broadcasts in the IPC channel. To scale the backend efficiently, we need to address the problem of emitting messages to all clients in the room. How can we prevent this and improve scalability, reducing IPC traffic for faster messaging through hash-based routing?
